src/whir/batching.rs

```diff
@@ -0,0 +1,819 @@
+    where
+        ProverFSState: FieldToUnitSerialize<F>
+            + BytesToUnitSerialize
+            + DigestToUnitSerialize<MerkleConfig>
+            + UnitToField<F>,
+    {
+        assert!(
+            polynomial_list.len() > 1,
+            "Batched evaluation requires more than one polynomial"
+        );
+
+        assert!(
+            polynomial_list.iter().all(|poly| {
+                poly.num_variables() == polynomial_list[0].num_variables()
+                    && poly.num_coeffs() == polynomial_list[0].num_coeffs()
+            }),
+            "Each polynomial must have same number of variables and same number of coefficients"
+        );
+
+        let batch_size = u32::try_from(polynomial_list.len()).unwrap().to_le_bytes();
+        prover_state.add_bytes(&batch_size)?;
+
+        let mut witness_list = Vec::new();
+
+        for poly in polynomial_list {
+            let individual_commitment = CommitmentWriter::commit(self, prover_state, poly.clone())?;
+            witness_list.push(individual_commitment);
+        }
+
+        let [batching_randomness] = prover_state.challenge_scalars()?;
+
+        Ok(BatchedWitness {
+            witness_list,
+            batching_randomness,
+        })
+    }
+}
+
+impl<F, MerkleConfig, PowStrategy> Prover<F, MerkleConfig, PowStrategy>
+where
+    F: FftField,
+    MerkleConfig: MTConfig<Leaf = [F]>,
+    PowStrategy: spongefish_pow::PowStrategy,
+{
+    fn to_multiroot_proof(
+        &self,
+        batched_proof: WhirProof<MerkleConfig, F>,
+        commitment: BatchedWitness<F, MerkleConfig>,
+    ) -> ProofResult<BatchedWhirProof<MerkleConfig, F>> {
+        assert!(
+            batched_proof.merkle_paths.len() > 0,
+            "There must be at least one merkle path"
+        );
+
+        let fold_sz = 1 << self.0.folding_factor.at_round(0);
+        let mut first_round_paths: Vec<(MultiPath<MerkleConfig>, Vec<Vec<F>>)> =
+            Vec::with_capacity(commitment.witness_list.len());
+
+        let ((batched_mt, batched_stir), rest) =
+            batched_proof.merkle_paths.as_slice().split_first().unwrap();
+        let stir_indexes = batched_mt.leaf_indexes.clone();
+
+        for merkle_tree in commitment.witness_list {
+            let paths = merkle_tree
+                .merkle_tree
+                .generate_multi_proof(stir_indexes.clone())
+                .map_err(|_| ProofError::InvalidProof)?;
+
+            let answers: Vec<_> = stir_indexes
+                .iter()
+                .map(|i| merkle_tree.merkle_leaves[i * fold_sz..(i + 1) * fold_sz].to_vec())
+                .collect();
+
+            first_round_paths.push((paths, answers));
+        }
+
+        assert!(
+            validate_stir_queries(
+                &commitment.batching_randomness,
+                &batched_stir,
+                &first_round_paths,
+            ),
+            "The computed value of STIR queries must match the value in the leaf nodes"
+        );
```

Comment:
Use a `debug_assert!` here and remove the conditional compilation on `validate_stir_queries`.